### PR TITLE
#187 Fix Segmentation + #151 Training Data Shape

### PR DESCRIPTION
**Pull request by WGierke**
The problem was that we haven't had a standardized way to handle training data. Now, the scans and binary nodule images have the shape 512x512x1024 so all DICOM images that have been rescaled to voxels fit into that. However, since this is a giant input matrix which can easily blow up the memory when training a neural network I added a `SegmentationModel` wrapper that takes this shape as an input and also predicts this shape, but the models that implement this interface can also rescale / crop / ... the input data as long as the predicted output is again of size 512x512x1024. For example consider the [`Simple3DModel`](https://github.com/concept-to-clinic/concept-to-clinic/compare/master...WGierke:187_fix_segmentation?expand=1#diff-7ebde87c389dab64f0c9e8a81e6b6391R49): it scales the input by 1/4 for each axis, learns a model based on that and after predicting it scales the predicted binary nodule mask again by factor 4 such that the output shape equals the given training data shape again. Note that I mainly moved the code from `models.py` to the classes in the `models/` directory to accomplish this.

## Reference to official issue
This references #187 and #151 since I wasn't able to cleanly separate them.

## How Has This Been Tested?
I added the piece of code that threw an error given in #187 as a test. All tests pass.

## CLA
- [X] I have signed the CLA; if other committers are in the commit history, they have signed the CLA as well

### CODE REVIEW

**Comment by lamby** [comment]
> can easily blow up the memory when training a neural network

Oh neat. How much was it using OOI?

**Comment by WGierke** [comment]
It would take 1 GB to connect all the input neurons to just 1 output neuron. Using max pooling and convolutional layers would of course crunch the number of trained weights. However, even the [2nd place of the NDSB 2017](http://juliandewit.github.io/kaggle-ndsb2017/) only used an input shape of 64x64x64 mm³ at max (which needed a Tesla K80 with 12 GB to train).

**Comment by reubano** [comment]
Nice work! Looking over it now.

File: prediction/src/algorithms/segment/src/data_generation.py [hunk 1]
```diff
@@ -3,22 +3,22 @@
...
+    lidc_idri_directories = [os.path.join(Config.FULL_DICOM_PATHS, name) for name in os.listdir(Config.FULL_DICOM_PATHS)
+                             if os.path.isdir(os.path.join(Config.FULL_DICOM_PATHS, name))]
 
-    for path in Config.FULL_DICOM_PATHS:
-        directories = path.split(os.path.sep)
-        lidc_id_path_index = 2 if in_docker else 5
-        lidc_id = directories[lidc_id_path_index]
+    for path in tqdm(lidc_idri_directories):
+        lidc_dir = path.split(IMAGES_FULL_DIR)[1]
+        lidc_id = lidc_dir.split('/')[1]
```
Comment (WGierke) [inline prediction/src/algorithms/segment/src/data_generation.py]: I'll change the slash to os.path.sep, sorry

File: prediction/src/algorithms/segment/src/model.py [hunk 2]
```diff
@@ -1,144 +0,0 @@
...
-
-        return Deconvolution3D(filters=nb_filters, kernel_size=kernel_size,
-                               output_shape=compute_level_output_shape(filters=nb_filters, depth=depth,
-                                                                       pool_size=pool_size, image_shape=image_shape),
-                               strides=strides, input_shape=compute_level_output_shape(filters=nb_filters,
-                                                                                       depth=depth + 1,
-                                                                                       pool_size=pool_size,
-                                                                                       image_shape=image_shape))
-    else:
-        return UpSampling3D(size=pool_size)
```
[the remark this answers is not in the document]
Comment (WGierke) [reply prediction/src/algorithms/segment/src/model.py]: Sure! I just thought "since it will be squashed anyways ...". But you're right, it doesn't make it easier for the reviewers.

**Comment by reubano** [comment]
Good to go! Mind squashing now?
